core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -2131,15 +2133,29 @@ public Optional<NodeRequirements> getNodeRequirements(int partitionId)
 
         public Optional<OutputDataSizeEstimateResult> getOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup, boolean parentEager)
         {
-            if (stage.getState() == StageState.FINISHED) {
-                return Optional.of(new OutputDataSizeEstimateResult(
-                        new OutputDataSizeEstimate(ImmutableLongArray.copyOf(outputDataSize)), OutputDataSizeEstimateStatus.FINISHED));
-            }
-            Optional<OutputDataSizeEstimateResult> result = getEstimatedOutputDataSize().or(() -> getEstimatedSmallStageOutputDataSize(stageExecutionLookup));
-            if (result.isEmpty() && parentEager) {
-                result = getEstimatedStageOutputSizeForEagerParent();
+            List<Supplier<Optional<OutputDataSizeEstimateResult>>> estimators = ImmutableList.of(
+                    () -> {
+                        if (stage.getState() != FINISHED) {
+                            return Optional.empty();
+                        }
+                        return Optional.of(new OutputDataSizeEstimateResult(new OutputDataSizeEstimate(ImmutableLongArray.copyOf(outputDataSize)), OutputDataSizeEstimateStatus.FINISHED));
+                    },
+                    this::getEstimatedOutputDataSize,
+                    () -> getEstimatedSmallStageOutputDataSize(stageExecutionLookup),
+                    () -> {
+                        if (!parentEager) {
+                            return Optional.empty();
+                        }
+                        return getEstimatedStageOutputSizeForEagerParent();
+                    });
+
+            for (Supplier<Optional<OutputDataSizeEstimateResult>> estimator : estimators) {
```

Comment:
This can also be written as a chain of `or()` calls 

like https://github.com/trinodb/trino/blob/3a8eaee6848e33a441b96c5731a7a8c2c78e3d2c/core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java#L596-L601

Reply:
It was chain of ors to some extent before - I did not like it.